amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala

```diff
@@ -6,100 +6,210 @@ package akka.stream.alpakka.amqp
+    ReusableAmqpConnectionProvider(provider: AmqpConnectionProvider, automaticRelease)
+
+  private sealed trait State
+  private case object Empty extends State
+  private case object Connecting extends State
+  private final case class Connected(connection: Connection, clients: Int) extends State
+  private case object Closing extends State
 }
 
-final case class ReusableAmqpConnectionProvider(settings: AmqpConnectionSettings, automaticRelease: Boolean = true)
+final case class ReusableAmqpConnectionProvider(provider: AmqpConnectionProvider, automaticRelease: Boolean = true)
     extends AmqpConnectionProvider {
-  var clients = 0
-  private val cachedConnectionRef = new AtomicReference[Option[Connection]](None)
-
-  override def get =
-    cachedConnectionRef
-      .updateAndGet(
-        asJavaUnaryOperator(
-          cachedConnection =>
-            cachedConnection match {
-              case Some(connection) => {
-                if (connection.isOpen) {
-                  clients += 1
-                  cachedConnection
-                } else {
-                  clients = 1
-                  Some(settings.getConnection)
-                }
-              }
-              case None => {
-                clients = 1
-                Some(settings.getConnection)
-              }
-          }
-        )
-      )
-      .get
-
-  override def release(connection: Connection) =
-    cachedConnectionRef
-      .updateAndGet(
-        asJavaUnaryOperator(
-          cachedConnection =>
-            cachedConnection match {
-              case Some(conn) =>
-                if (conn != connection)
-                  throw new IllegalArgumentException("Can't release a connection that's not owned by this provider")
-                if (!automaticRelease) {
-                  clients = 0
-                  if (connection.isOpen) connection.close()
-                  None
-                } else {
-                  clients -= 1
-                  if (clients == 0) {
-                    if (connection.isOpen) connection.close()
-                    None
-                  }
-                  cachedConnection
-                }
-              case None => None
-          }
-        )
-      )
+
+  import akka.stream.alpakka.amqp.ReusableAmqpConnectionProvider._
+  private val state = new AtomicReference[State](Empty)
+
+  @tailrec
+  override def get: Connection = state.get match {
+    case Empty =>
+      if (state.compareAndSet(Empty, Connecting)) {
+        val connection = provider.get
+        state.compareAndSet(Connecting, Connected(connection, 0))
+        connection
+      } else get
+    case c @ Connected(connection, clients) =>
+      if (state.compareAndSet(c, Connected(connection, clients + 1))) connection
+      else get
+    case _ => get
+  }
+
+  @tailrec
+  override def release(connection: Connection): Unit = state.get match {
+    case c @ Connected(cachedConnection, clients) =>
+      if (state.compareAndSet(c, Closing)) {
```

Comment:
Check the cached connection as the first thing in `release` instead of modifying the state and then checking.

Don't change to closing unless the clients are 0 and automatic release is true, for other cases it should just subtract one from the client count.